tests/core/util/test_db_wrapper.py

```diff
@@ -0,0 +1,221 @@
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    argnames="acquire_outside",
+    argvalues=[pytest.param(False, id="not acquired outside"), pytest.param(True, id="acquired outside")],
+)
+async def test_concurrent_writers(acquire_outside: bool) -> None:
+
+    async with DBConnection(2) as db_wrapper:
+        await setup_table(db_wrapper)
+
+        concurrent_task_count = 200
+
+        async with contextlib.AsyncExitStack() as exit_stack:
+            if acquire_outside:
+                await exit_stack.enter_async_context(db_wrapper.write_db())
+
+            tasks = []
+            for index in range(concurrent_task_count):
+                task = asyncio.create_task(increment_counter(db_wrapper))
+                tasks.append(task)
+
+        await asyncio.wait_for(asyncio.gather(*tasks), timeout=None)
+
+        async with db_wrapper.read_db() as connection:
+            async with connection.execute("SELECT value FROM counter") as cursor:
+                row = await cursor.fetchone()
+
+            assert row is not None
+            [value] = row
+
+    assert value == concurrent_task_count
+
+
+@pytest.mark.asyncio
+async def test_writers_nests() -> None:
+    async with DBConnection(2) as db_wrapper:
+        await setup_table(db_wrapper)
+        async with db_wrapper.write_db() as conn1:
+            async with conn1.execute("SELECT value FROM counter") as cursor:
+                value = await get_value(cursor)
+            async with db_wrapper.write_db() as conn2:
+                assert conn1 == conn2
+                value += 1
+                await conn2.execute("UPDATE counter SET value = :value", {"value": value})
+                async with db_wrapper.write_db() as conn3:
+                    assert conn1 == conn3
+                    async with conn3.execute("SELECT value FROM counter") as cursor:
+                        value = await get_value(cursor)
+
+    assert value == 1
+
+
+@pytest.mark.asyncio
+async def test_partial_failure() -> None:
+    async with DBConnection(2) as db_wrapper:
+        await setup_table(db_wrapper)
+        async with db_wrapper.write_db() as conn1:
+            await conn1.execute("UPDATE counter SET value = 42")
+            async with conn1.execute("SELECT value FROM counter") as cursor:
+                assert await get_value(cursor) == 42
+            try:
+                async with db_wrapper.write_db() as conn2:
+                    await conn2.execute("UPDATE counter SET value = 1337")
+                    async with conn1.execute("SELECT value FROM counter") as cursor:
+                        assert await get_value(cursor) == 1337
+                    raise RuntimeError("failure within a sub-transaction")
+            except BaseException:
+                pass
+            async with conn1.execute("SELECT value FROM counter") as cursor:
+                value = await get_value(cursor)
+
+    # the write of 1337 failed, and was restored to 42
+    assert value == 42
```

Comment:
I'm somewhat tempted to in general collect the values instead of asserting then assert at the end.  For example, with the existing `except BaseException: pass`, any non-syntax coding error could be made inside the `try:` and the test would still pass without having checked what it meant to.  Though, checking coverage in tests would catch some such cases.  None the less, sometimes a useful pattern to make tests less likely to false pass.

Reply:
you're right. I'll fix this